VSCODE-72: create a new runtime instance for each evaluation

alenakhineika opened the pull request:
<!-- Ticket number and a general summary of your changes in the Title above -->
<!-- e.g. COMPASS-1111: updates ace editor width in agg pipeline view -->

<!--- The following fields are not obligatory. Use your best judgement on what you think is applicable to the work you've done -->

## Description
<!--- Describe your changes in detail -->
<!--- If applicable, describe (or illustrate) architecture flow -->
1) Use a new version of `mongosh` with https://github.com/mongodb-js/mongosh/pull/64 merged that fixes missing db interaction commands in browser runtime (eg. `use('db')`).

2) Create a new runtime instance for each run of the playground. It will allow to repeatedly run the same commands and avoid errors such as `Unable to run playground: Identifier ‘s’ has already been declared`.

Example from @mmarcon:

```
const s = 3, l = 1;
db.users.find().skip(s).limit(l);
```

_"When I run this the first time, this works well. But then if I re-run it, I get “Unable to run playground: Identifier ‘s’ has already been declared” which in the REPL makes sense but in the playground case I’d expect that every run starts with a clean context."_

3) Display active connection instead of active db. When switching connections update active connection name in code lenses.

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it's updating a dependancy, link to the Pull Request that originally introduced the fix -->
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

Anemy commented on src/editors/playgroundController.ts:
```diff
@@ -17,9 +18,8 @@ const log = createLogger('editors controller');
 export default class PlaygroundController {
   _context?: vscode.ExtensionContext;
```
Since we always assign this is the constructor I think we can lose the `?` (same with `_connectionController` and `_activeConnectionCodeLensProvider`) and that should allow us to get rid of an extra check or two.

Anemy commented on src/editors/playgroundController.ts:
```diff
@@ -125,4 +119,15 @@ export default class PlaygroundController {
+      this._connectionController.removeEventListener(
+        DataServiceEventTypes.ACTIVE_CONNECTION_CHANGED,
+        () => {
+          this.showActiveConnectionInPlayground('Playground disconnected');
```
I'm not sure I follow the expected behavior here, when the mongodb app deactivates we want to show this message?

alenakhineika replied:
Yeah, I just added it to check when deactivation happens, but have never seen this message actually. What do you think is better here? Empty function?

Anemy replied:
I think empty function is probably better just so we preserve what is in the `output`.

Anemy commented on src/templates/playgroundTemplate.ts:
```diff
@@ -1,12 +1,15 @@
 ];
-db.my_collection.aggregate(agg);
+
+db.myCollection.aggregate(agg);
```
👍

Anemy commented on src/test/suite/editors/playgroundController.test.ts:
```diff
@@ -49,5 +54,43 @@ suite('Playground Controller Test Suite', () => {
+      expect(await testPlaygroundController.evaluate(`
+        var x = 1;
+        x + 2
+      `)).to.be.equal('3');
```
yesss! scripting!

Anemy commented on src/editors/activeConnectionCodeLensProvider.ts:
```diff
@@ -1,10 +1,9 @@
   private _connectionController: any;
-  private _runtime: any;
-  private _activeDB?: string;
+  private _activeConnectionName?: string;
```
If a playground is opened or created before a connection is established, would this be `undefined`? I think another way we could do this which might allow us not to keep the connection name here would be to call `this._connectionController.getActiveConnectionName()` in the method `resolveCodeLens` for the connection name, and then we can handle the case where there is no active connection there. It also might allow us not to have the event listener `ACTIVE_CONNECTION_CHANGED` altogether.

alenakhineika replied:
If a playground is opened or created before a connection is established, it will be no code lenses at all. So we don't want to show an active connection name when there is no connection. In case if we get rid of ACTIVE_CONNECTION_CHANGED listener we won't be able to update code lenses with active connection state if we didn't change a state of a playground. I mean we need to type something to invoke code lenses update. And I wanted to change the name in real-time.  Pls, try to play with a playground and let me think how does it feel like it works now.

Anemy replied:
Ah gotcha. It looks like `ACTIVE_CONNECTION_CHANGED` is definitely a useful event to have. I think we can avoid storing the `_activeConnectionName` on the code lens though if we get the name from `this._connectionController.getActiveConnectionName()` in the `resolveCodeLens` method, since the code lens is resolved everytime `this._onDidChangeCodeLenses.fire()` is called. I'd like to avoid duplicating the storage of the connection name if possible, tried it on local and it seemed to do the trick.

mmarcon commented on src/templates/playgroundTemplate.ts:
```diff
@@ -1,12 +1,15 @@
+
+db.myCollection.aggregate(agg);
 `;
 
```
@Anemy had a preference for closing the sentences in the comments with a `.`. I think it makes sense, although I don't have a strong opinion either way.

alenakhineika replied:
I don't have a strong opinion either, I prefer without, but I will return them back if having them makes sense for both of you.

mmarcon commented on src/editors/activeConnectionCodeLensProvider.ts:
```diff
@@ -37,9 +36,12 @@ export default class ActiveDBCodeLensProvider implements vscode.CodeLensProvider
   }
 
   public resolveCodeLens?(codeLens: vscode.CodeLens): vscode.CodeLens {
+    const message = `Active connection is ${this._activeConnectionName}`;
```
👍

alenakhineika commented:
Thanks for the awesome recommendations in the review! So what I did:

- Returned `.` to the template 😏
- Renamed code lenses text prefix
- Removed the message after removing the listener
- Added a test that runs the playground two times in a raw and checks that the second result is correct. That covers the use case form the PR description
- Removed extra checks for context and connection controller
- Removed the connection name property from code lenses class

Anemy commented:
lgtm!
